File: app-laravel/src/Shared/Application/Middleware/LoggingMiddleware.php
<?php

declare(strict_types=1);

namespace Src\Shared\Application\Middleware;

use Illuminate\Support\Facades\Log;
use Src\Shared\Application\Bus\Command;

/**
 * LOGGING MIDDLEWARE
 * =================
 * Hər Command-ın başlanğıc və bitmə vaxtını log edir.
 * Debug və monitoring üçün çox faydalıdır.
 *
 * Pipeline-da yeri:
 * Command → [LOGGING] → [Validation] → [Transaction] → Handler
 */
class LoggingMiddleware implements Middleware
{
    public function handle(Command $command, callable $next): mixed
    {
        $commandName = get_class($command);

        // Command başlayanda log
        Log::info("Command başladı: {$commandName}", [
            'command' => get_object_vars($command),
            'timestamp' => now()->toISOString(),
        ]);

        $startTime = microtime(true);

        try {
            // Növbəti middleware-ə ötür (və ya Handler-ə)
            $result = $next($command);

            $duration = round((microtime(true) - $startTime) * 1000, 2);

            Log::info("Command uğurla tamamlandı: {$commandName}", [
                'duration_ms' => $duration,
            ]);

            return $result;
        } catch (\Throwable $e) {
            $duration = round((microtime(true) - $startTime) * 1000, 2);

            Log::error("Command xəta ilə bitdi: {$commandName}", [
                'duration_ms' => $duration,
                'error' => $e->getMessage(),
            ]);

            throw $e;
        }
    }
}

File: app/spring/src/main/java/az/ecommerce/shared/application/middleware/LoggingMiddleware.java
package az.ecommerce.shared.application.middleware;

import az.ecommerce.shared.application.bus.Command;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

/**
 * Pipeline mövqeyi: 1-ci (ən xarici).
 * Hər command-ı və nəticəsini loga yazır.
 * Laravel: LoggingMiddleware.php
 */
@Component
public class LoggingMiddleware implements CommandMiddleware {

    private static final Logger log = LoggerFactory.getLogger(LoggingMiddleware.class);

    @Override
    public <R> R handle(Command<R> command, CommandPipeline<R> next) {
        long start = System.currentTimeMillis();
        log.info("[CMD] {} başladı", command.getClass().getSimpleName());
        try {
            R result = next.proceed(command);
            log.info("[CMD] {} bitdi ({}ms)",
                    command.getClass().getSimpleName(), System.currentTimeMillis() - start);
            return result;
        } catch (Exception ex) {
            log.error("[CMD] {} XƏTA: {}", command.getClass().getSimpleName(), ex.getMessage(), ex);
            throw ex;
        }
    }

    @Override
    public int order() {
        return 10;
    }
}
